lib/service/connect.go

```diff
@@ -684,6 +687,112 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
 	return connector, nil
 }
 
+func (process *TeleportProcess) initOpenSSH() {
+	process.RegisterWithAuthServer(types.RoleNode, SSHIdentityEvent)
+	process.SSHD = openssh.NewSSHD(process.Config.OpenSSH.RestartCommand)
+	process.RegisterCriticalFunc("openssh.rotate", process.syncOpenSSHRotationState)
+}
+
+func (process *TeleportProcess) syncOpenSSHRotationState() error {
+	if _, err := process.WaitForEvent(process.GracefulExitContext(), TeleportReadyEvent); err != nil {
+		return trace.Wrap(err)
+	}
+	conn, err := process.WaitForConnector(SSHIdentityEvent, nil)
+	if conn == nil {
+		return trace.Wrap(err)
+	}
+	defer conn.Close()
+
+	status, err := process.syncRotationState(conn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	id, err := process.storage.ReadIdentity(auth.IdentityCurrent, types.RoleNode)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	ctx := process.ExitContext()
+	cas, err := conn.Client.GetCertAuthorities(ctx, types.OpenSSHCA, false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	keysDir := process.Config.OpenSSH.KeysDir
+	if err := openssh.WriteKeys(keysDir, id, cas); err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = process.SSHD.UpdateConfig(openssh.SSHDConfigUpdate{
+		SSHDConfigPath:         process.Config.OpenSSH.SSHDConfigPath,
+		TeleportSSHDConfigPath: openssh.SSHDConfigPath,
+		OpenSSHCAPath:          filepath.Join(keysDir, openssh.TeleportOpenSSHCA),
+		HostKeyPath:            filepath.Join(keysDir, openssh.TeleportKey),
+		HostCertPath:           filepath.Join(keysDir, openssh.TeleportCert),
+	}, process.Config.OpenSSH.RestartSSHD)
+
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	state, err := process.storage.GetState(types.RoleNode)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	mostRecentRotation := state.Spec.Rotation
+	if !status.needsReload {
```

Comment:
Even if we got `needsReload` because the host CA was rotated we might still have an even more recent openssh CA, right?